### PR TITLE
*: support time based auto compaction.

### PR DESCRIPTION
**Pull request by xiang90**
Fix https://github.com/coreos/etcd/issues/3906.

We will have extensive doc to talk about what is compaction
and what is auto compaction soon.

### CODE REVIEW

File: compactor/compactor.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+type Compactable interface {
+	Compact(ctx context.Context, r *pb.CompactionRequest) (*pb.CompactionResponse, error)
+}
+
+type RevGetter interface {
+	Rev() int64
+}
+
+type TimeBased struct {
```
Comment (heyitsanthony) [inline compactor/compactor.go]: `Periodic`?

File: compactor/compactor.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	clock        clockwork.Clock
+	periodInHour int
```
Comment (gyuho) [inline compactor/compactor.go]: so if `periodInHour` is 3, it auto-compacts every 3-hour?

**Comment by heyitsanthony** [comment]
It might be a useful to have a policy that won't compact past the minimum revision for open watchers, but that can be another PR

**Comment by xiang90** [comment]
This is the simplest approach. I tried to implement it in a bunch of other ways to persist the time-rev map across the cluster consistently and feel they are overkilled. The current guarantees is that as long as the cluster has quorum or the members are all alive, the compaction will be started in time. If the whole cluster is down, the compaction might not be delivered in time since a restart will reset the compaction timer.

File: compactor/compactor.go (hunk 2 above)
Comment (xiang90) [inline compactor/compactor.go]: it compacts every 3 hours and keeps the most recent 3 hour history.

Comment (gyuho) [inline compactor/compactor.go]: ok got it. thanks.

File: etcdserver/raft.go [hunk 3]
```diff
@@ -159,10 +159,16 @@ func (r *raftNode) start(s *EtcdServer) {
...
 						}
+						if r.s.compactor != nil {
+							r.s.compactor.Resume()
+						}
 					} else {
 						if r.s.lessor != nil {
 							r.s.lessor.Demote()
 						}
+						if r.s.compactor != nil {
+							r.s.compactor.Pause()
```
Comment (gyuho) [inline etcdserver/raft.go]: could you explain why we pause compaction here?

File: compactor/compactor.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	t.paused = true
+}
+
+func (t *TimeBased) Resume() {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.paused = false
+}
+
+func (t *TimeBased) getRev(h int) int64 {
```
Comment (heyitsanthony) [inline compactor/compactor.go]: what does `h` do?

File: etcdserver/raft.go (hunk 3 above)
Comment (xiang90) [inline etcdserver/raft.go]: we want the leader to control the compaction. the follower could help though. (for example, the leader might be a restarted node and just reset the timer and fail to compact in time.) but i feel this is somewhat overly complicated the issue. i want to see how the simplest solution goes before making this more advanced.

File: compactor/compactor.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+				return
+			case <-clock.After(checkCompactionInterval):
+				t.mu.Lock()
+				p := t.paused
+				t.mu.Unlock()
+				if p {
+					continue
+				}
+			}
+			if clock.Now().Sub(last) >= time.Duration(t.periodInHour)*time.Hour {
```
Comment (heyitsanthony) [inline compactor/compactor.go]: `if clock.Now().Sub(last) < time.Duration(t.periodInHour)*time.Hour { continue }`?

File: etcdserver/raft.go (hunk 3 above)
Comment (gyuho) [inline etcdserver/raft.go]: got it. yeah I think making the leader do compaction simplifies for the first stage. thanks for explanation.

File: compactor/compactor.go (hunk 4 above)
Comment (xiang90) [inline compactor/compactor.go]: fixed.

**Comment by heyitsanthony** [comment]
lgtm

**Comment by xiang90** [comment]
@heyitsanthony All fixed.

**Pull request merged**